proposals/0000-namespace-specified-imports.rst

```diff
@@ -0,0 +1,554 @@
+
+  import Data.Functor.Product ( Product ( type Pair ) )
+
+Given that this is essentially a bug, it seems unlikely that user code relies
+on it. Moreover it can be fixed by simply removing the bogus ``type``
+namespace specifier. Thus we do not expect significant breakage from this
+change.
+
+Existing code that uses the ``pattern`` keyword (with ``PatternSynonyms``) in
+import/export lists and uses ``-Wcompat`` (or eventually ``-Wall``) will
+receive a warning from ``-Wpattern-namespace-specifier`` until it migrates to
+use ``data`` instead (with ``ExplicitNamespaces``).  In the future we may wish
+to remove support for ``pattern`` as a namespace specifier entirely, but doing
+so is not part of this proposal, so this is not a breaking change.
+
+
+Alternatives
+------------
+
+* We could use ``value``, ``term``, ``pattern``, or any other keyword instead of
+  ``data`` to denote the data namespace.  It seems preferable to use ``data`` as
+  it is an existing keyword (unlike ``value`` and ``term``) and unlike
+  ``pattern`` it more clearly refers to the data namespace.  However this may
+  lead to beginner confusion if expressions like ``import M (data f)`` are used,
+  since ``data`` refers to the namespace rather than a datatype.
+
+* There are alternatives to the import syntax proposed here, for example
+  `proposal #340 <https://github.com/ghc-proposals/ghc-proposals/pull/340>`_
+  proposes ``import M as (T, D)`` syntax. More examples of alternative syntax:
+
+  - ``import M as {T, D}``
+  - ``import M type as T data as D``
+  - ``import M as (type T, data D)``
+
+* Instead of introducing new syntax we could use the existing syntax for
+  explicit import lists:
+
+  .. code:: haskell
+
+     import Data.Proxy( Proxy ) qualified as T
+     import Data.Proxy( pattern Proxy ) qualified as D
+
+  However it is significantly less convenient: you can’t import all the
+  things at once without manually listing every single one of them.
+
+* This proposal does not introduce a way to disambiguate namespaces at use
+  sites, corresponding to the prefix ``'`` syntax used by ``DataKinds`` and the
+  prefix ``'`` and ``''`` used by ``TemplateHaskell``.  Instead, disambiguation
+  must happen at the level of imports, and cannot be used for definitions within
+  the same module. `Proposal #214
+  <https://github.com/ghc-proposals/ghc-proposals/pull/214>`_ suggested using
+  prefix ``data.`` and ``type.`` systematically for this purpose.  However, that
+  would be syntactically noisy and there was difficulty gathering consensus for
+  that approach.
+
+* Rather than modifying ``ExplicitNamespaces`` to allow namespace-specified
+  imports, we could introduce a new extension ``NamespacedImports``.  This would
+  make it clearer whether code depends on the new feature, or whether the older
+  version of ``ExplicitNamespaces`` was enough.  However the general consensus
+  seems to favour reducing the number of extensions over avoiding change to
```

Comment:
That's debatable :-). We'll know more once we have concluded the ongoing policy discussion. But since we didn't add ExplicitNamespaces to GHC2023, likely fine here